### PR TITLE
backend/feat/3615 : modified the cash collection api

### PR DESCRIPTION
**Pull request by AshishSaini25**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [x] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

## Description

modified the cash collection api
### Additional Changes

- [ ] This PR modifies the database schema (database migration added)
- [ ] This PR modifies dhall configs/environment variables
- [ ] This PR contains API breaking changes

## Checklist

- [ ] I formatted the code and addressed linter errors `./dev/format-all-files.sh`
- [ ] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable
- [ ] No leak detected in leakcanary

### CODE REVIEW

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Dashboard/Revenue.hs [hunk 1]
```diff
@@ -78,31 +86,33 @@ getCollectionHistory merchantShortId volunteerId place mbFrom mbTo = do
...
           | tmp <-
-              DL.groupBy (\(_, a) (_, b) -> a == b) $
+              DL.groupBy (\(_, a) (_, b) -> makeGranular a == makeGranular b) $
                 sortBy (comparing snd) [(fee.numRides, fee.collectedAt) | fee <- collections]
         ]
-      driversPmtTs = [(length tmp, listToMaybe tmp) | tmp <- DL.group $ DL.sort $ catMaybes [fee.collectedAt | fee <- collections]]
+      driversPmtTs = [(length tmp, listToMaybe (catMaybes tmp)) | tmp <- DL.groupBy (\a b -> makeGranular a == makeGranular b) $ DL.sort [fee.collectedAt | fee <- collections]]
   Common.CollectionListRes totalCnt totalFeeCollected totalRides totalDrivers collectionsTs numRidesTs driversPmtTs
   where
+    makeGranular (Just timeVal) = makeSlidingWindowKey granuality "" timeVal
```
Comment (0utkarsh) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Dashboard/Revenue.hs:114]: Isn't this still on timestamp basis always?

**Review by 0utkarsh — CHANGES_REQUESTED** [review]

File: Backend/app/dashboard/CommonAPIs/src/Dashboard/ProviderPlatform/Revenue.hs [hunk 2]
```diff
@@ -83,3 +88,8 @@ data DriverFeeAPIEntity = DriverFeeAPIEntity
     collectedAt :: Maybe UTCTime
   }
   deriving (Generic, Show, FromJSON, ToJSON, ToSchema)
+
+data Basis = Daily | Hourly
```
[the remark this answers is not in the document]
Comment (AshishSaini25) [reply Backend/app/dashboard/CommonAPIs/src/Dashboard/ProviderPlatform/Revenue.hs:92]: like this is suggested by @hemantmangla  in the ticket.
am i going in correct direction ??
https://github.com/nammayatri/nammayatri/issues/3615#issuecomment-1759404747

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Dashboard/Revenue.hs (hunk 1 above)
Comment (AshishSaini25) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Dashboard/Revenue.hs:114]: no like if the param is hourly then this function will group on the basis of hours , and same for daily.

File: Backend/app/dashboard/CommonAPIs/src/Dashboard/ProviderPlatform/Revenue.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (0utkarsh) [reply Backend/app/dashboard/CommonAPIs/src/Dashboard/ProviderPlatform/Revenue.hs:92]: Yeah, but I don't see any reason to remove the timestamp case. This is what you suggested @hkmangla?

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Dashboard/Revenue.hs [hunk 3]
```diff
@@ -61,13 +64,18 @@ getAllDriverFeeHistory merchantShortId mbFrom mbTo = do
...
   now <- getCurrentTime
   merchant <- findMerchantByShortId merchantShortId
   let defaultFrom = UTCTime (fromGregorian 2020 1 1) 0
       from_ = fromMaybe defaultFrom mbFrom
       to = fromMaybe now mbTo
+  let rangeBasis = case basis of
+        Just Common.Daily -> KS.Days
+        Just Common.Hourly -> KS.Hours
+        Nothing -> KS.Days
+  when (rangeBasis == KS.Hours && diffUTCTime to from_ > 24 * 60 * 60) $ throwError $ InvalidRequest "Hourly basis data can be seen only in the range of one day "
```
Comment (0utkarsh) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Dashboard/Revenue.hs:78]: I think we can remove the query param and just show on hourly basis if its a day and daily basis if more than one day. Can you make that change? I'll merge the PR after that

**Review by 0utkarsh — CHANGES_REQUESTED** [review]

**Review by 0utkarsh — APPROVED** [review]

**Review by hkmangla — APPROVED** [review]

**Pull request merged**